### PR TITLE
(SIMP-2911) Replace 'classes' with 'simp::classes' and refactor scenarios to be in-module data

### PR DESCRIPTION
**Pull request by trevor-vaughan**
* Remove support for the legacy 'classes' array from the entire
      SIMP stack.
    * Add EL6 specific class items from the stock hierarchy
    * Refactor to use data-in-modules
    * Move classification out of simp::scenario::* into data/common.yaml and
      data/os/RedHat-6.yaml
    * Add simp::scenario_map, simp::server::scenario_map for the lookup keys
    * Add simp::classes and simp::server::classes which allow for the admin
      to specify site or host specific class overrides from the scenario list,
      using '--' as the knockout prefix
    * Add a function 'simp::knockout' that condenses arrays using -- as the
      knockout_prefix in a reliable manner.
    * Add a function 'simp::classifier' that uses a hash table to classify a
      node using the scenarios as the default list, and then merging in a set
      of overrides
    * Add very verbose regression tests to validate the conversion.

SIMP-2911 #close

### CODE REVIEW

File: manifests/init.pp [hunk 1]
```diff
@@ -130,7 +114,7 @@
 class simp (
   Enum['simp','simp_lite','poss'] $scenario                   = 'simp',
   Boolean                         $enable_data_includes       = true,
-  String                          $data_includes_array_name   = 'classes',
+  Optional[Array]                 $classes                    = lookup('simp::classes', Optional[Array], { 'strategy' => 'deep', 'knockout_prefix' => '--' }, undef),
```
Comment (lamawithonel) [inline manifests/init.pp]: Have you tested this? Last I knew "knockout_prefix" didn't work with arrays.

File: manifests/init.pp [hunk 2]
```diff
@@ -130,7 +114,7 @@
 class simp (
   Enum['simp','simp_lite','poss'] $scenario                   = 'simp',
   Boolean                         $enable_data_includes       = true,
```
Comment (lamawithonel) [inline manifests/init.pp]: Based on the change to the logic below, `$enable_data_includes` should be removed.

File: manifests/scenario/simp.pp [hunk 3]
```diff
@@ -65,4 +65,10 @@
   include 'sudosh'
   # Ensure that only services that are defined in Puppet are going to be enabled and run
   include 'svckill'
+
+  # EL6 only material
+  if ($facts['os']['name'] in ['RedHat', 'CentOS']) and (versioncmp($facts['os']['release']['major'], '6') == 0) {
+    include 'acpid'
+    include 'upstart'
+  }
```
Comment (lamawithonel) [inline manifests/scenario/simp.pp]: This needs a comment in the changelog and commit message, or it should be moved to a separate PR.

File: manifests/init.pp (hunk 1 above)
Comment (trevor-vaughan) [inline manifests/init.pp]: @lamawithonel Yeah, this needs to be pushed down into the data layer so that it can be overridden later.

Comment (trevor-vaughan) [inline manifests/init.pp]: Oh, the knockout prefix does work with arrays but you have to take care that people don't mess it up and leave things with '--' in them in the final result.

**Comment by trevor-vaughan** [comment]
This will probably be replaced with whatever @heliocentric is putting together

File: metadata.json [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "simp-simp",
-  "version": "3.1.0",
+  "version": "3.2.0",
```
Comment (heliocentric) [inline metadata.json]: agreed, I'll bump it. This is a good example of why having a split develop/master can be a good thing, gets you an unstable integration branch that can churn without bumping vers, particularly on ideas that aren't fully proven out yet.

File: manifests/scenario/poss.pp [hunk 5]
```diff
@@ -1,38 +0,0 @@
-# **NOTE: THIS IS A [PRIVATE](https://github.com/puppetlabs/puppetlabs-stdlib#assert_private) CLASS**
```
Comment (heliocentric) [inline manifests/scenario/poss.pp]: I need to add this in for the logic it contains, just not classify anything, as it's actually an api

File: manifests/server.pp [hunk 6]
```diff
@@ -29,36 +29,13 @@
...
-  include '::simp::server::rsync_shares'
-  include '::simp::sysctl'
+  Boolean $auditd          = simplib::lookup('simp_options::auditd', { 'default_value' => false }),
+  String $scenario         = $simp::scenario,
+  Array $classes           = [],
+  Hash $scenario_map,
+) inherits simp {
 
+  simp::classifier($scenario, $scenario_map, $classes)
   if $clamav  { include '::clamav' }
```
Comment (heliocentric) [inline manifests/server.pp]: so I don't forget, 6.1 might be a good chance to look at making simp_options::* toggle classes in simp::classifier. Just for the lulz

File: data/common.yaml [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    - ssh
+    - sudosh
+  simp:
+# Shared with poss
+    - pupmod
+# Shared with simp_lite
+    - simp::scenario::base
+    - aide
+    - auditd
+    - clamav
```
Comment (heliocentric) [inline data/common.yaml]: I'm told this is an accident and should be in simp::scenario::base and gated by booleans

**Comment by lamawithonel** [comment]
I put a PR in against this PR.  See trevor-vaughan/pupmod-simp-simp#1

File: functions/knockout.pp [hunk 8]
```diff
@@ -0,0 +1,14 @@
+function simp::knockout(Array $array) {
```
Comment (jeefberkey) [inline functions/knockout.pp]: Function header needed

File: manifests/init.pp [hunk 9]
```diff
@@ -128,9 +120,10 @@
 # @author Trevor Vaughan <tvaughan@onyxpoint.com>
 #
 class simp (
-  Enum['simp','simp_lite','poss'] $scenario                   = 'simp',
+  Hash                            $scenario_map,
```
Comment (jeefberkey) [inline manifests/init.pp]: What's this used for?

File: functions/merge_classes.pp [hunk 10]
```diff
@@ -0,0 +1,22 @@
+# simp::merge_classes()
+#
+# Includes the classes from the selected scenario, plus any additions from the
+# the classes parameter, and minus any classes prefixed with `--`.
```
Comment (lamawithonel) [inline functions/merge_classes.pp]: This needs to be updated. The function does not `include` classes.

File: functions/merge_classes.pp [hunk 11]
```diff
@@ -0,0 +1,22 @@
+# simp::merge_classes()
```
Comment (lamawithonel) [inline functions/merge_classes.pp]: Can this be `merge_scenario_classes()`? The scenario part seems more important.

File: functions/merge_classes.pp [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+#
+# Includes the classes from the selected scenario, plus any additions from the
+# the classes parameter, and minus any classes prefixed with `--`.
+#
+# @private
+function simp::merge_classes(Optional[Array] $classes, String $scenario = "", Hash[String, Array] $scenario_map = {}) {
```
Comment (lamawithonel) [inline functions/merge_classes.pp]: `$classes` should be at the end of the params list, and defaulted to `undef`.

File: functions/merge_classes.pp [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+  }
+  if ($scenario_map == {}) {
+    $selected_classes = $_classes
+  } elsif $scenario_map.has_key($scenario) {
+    $selected_classes = $scenario_map[$scenario] + $_classes
+  } else {
+    fail("ERROR - Invalid scenario '${scenario}' for the given scenario map.")
+  }
+
+  simp::knockout($selected_classes)
```
Comment (lamawithonel) [inline functions/merge_classes.pp]: Since this doesn't `include` anymore, I think this should happen outside the function, in the manifest.

File: spec/classes/server_spec.rb [hunk 14]
```diff
@@ -29,6 +29,85 @@
...
+            'poss' => {
+              'contains' => [
+                poss,
+                simp_lite,
+                simp,
+              ],
+              'does_not_contain' => [
+              ]
+            }
+          }
```
Comment (lamawithonel) [inline spec/classes/server_spec.rb]: Ignore this if it's too hard: This fixture is ugly and hard to follow.  Might it be better to put these in a YAML file and read them in?

Comment (heliocentric) [inline spec/classes/server_spec.rb]: ignored!

**Comment by trevor-vaughan** [comment]
@lamawithonel Please re-review or dismiss your negative review so we can get this done.

File: functions/merge_classes.pp (hunk 13 above)
Comment (lamawithonel) [inline functions/merge_classes.pp]: This is still there. There's no reason this function to depend on the other function. It's just extra coupling for no reason.

**Comment by lamawithonel** [comment]
This still need tests in `spec/functions/simp_scenario_merge_spec.rb`

**Comment by lamawithonel** [comment]
@nick-markowski Please put the XXX comment back in. If not for all of them, at least for one. "Why" is the most important part of any code comment.
